// @shreaysh/dpdp/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class DPDPClient {
    constructor(apiKey, baseUrl) {
        if (!apiKey)
            throw new Error('API key is required to initialize DPDPClient');
        this.apiKey = apiKey;
        this.baseUrl = baseUrl.endsWith('/') ? baseUrl.slice(0, -1) : baseUrl;
    }
    async request(endpoint, options = {}) {
        const url = `${this.baseUrl}${endpoint}`;
        // Merge headers
        const headers = {
            'Content-Type': 'application/json',
            'x-api-key': this.apiKey,
            ...options.headers,
        };
        const response = await fetch(url, { ...options, headers });
        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            throw new Error(errorData.error || `HTTP error! status: ${response.status}`);
        }
        return response.json();
    }
    async createUser() {
        const response = await this.request('/dpdp/users', {
            method: 'POST',
        });
        return response.userUID;
    }
    async storePermission(userUID, data) {
        const response = await this.request('/dpdp/permissions', {
            method: 'POST',
            body: JSON.stringify({ userUID, data }),
        });
        return response.permissionId;
    }
    async storeData(userUID, data) {
        const response = await this.request('/dpdp/data', {
            method: 'POST',
            body: JSON.stringify({ userUID, data }),
        });
        return response.dataId;
    }
    async revokePermission(userUID, id) {
        await this.request('/dpdp/permissions/revoke', {
            method: 'POST',
            body: JSON.stringify({ userUID, permissionId: id }),
        });
    }
    async revokeData(userUID, id) {
        await this.request('/dpdp/data/revoke', {
            method: 'POST',
            body: JSON.stringify({ userUID, dataId: id }),
        });
    }
    async deletePermission(userUID, id) {
        await this.request('/dpdp/permissions/delete', {
            method: 'DELETE',
            body: JSON.stringify({ userUID, permissionId: id }),
        });
    }
    async deleteData(userUID, id) {
        await this.request('/dpdp/data/delete', {
            method: 'DELETE',
            body: JSON.stringify({ userUID, dataId: id }),
        });
    }
    async viewPermission(userUID, id) {
        return this.request(`/dpdp/permissions/${userUID}/${id}`, {
            method: 'GET',
        });
    }
    async viewData(userUID, id) {
        return this.request(`/dpdp/data/${userUID}/${id}`, {
            method: 'GET',
        });
    }
}
exports.default = DPDPClient;
