// @shreaysh/dpdp/src/index.ts
export interface PermissionData {
    subject: string;
    description: string;
}

export interface DataRecordInput {
    type: 'email' | 'phone' | 'address';
    value: string;
}

export default class DPDPClient {
    private apiKey: string;
    private baseUrl: string;

    constructor(apiKey: string, baseUrl: string) {
        if (!apiKey) throw new Error('API key is required to initialize DPDPClient');
        this.apiKey = apiKey;
        this.baseUrl = baseUrl.endsWith('/') ? baseUrl.slice(0, -1) : baseUrl;
    }

    private async request<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
        const url = `${this.baseUrl}${endpoint}`;

        // Merge headers
        const headers = {
            'Content-Type': 'application/json',
            'x-api-key': this.apiKey,
            ...options.headers,
        };

        const response = await fetch(url, { ...options, headers });

        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            throw new Error(errorData.error || `HTTP error! status: ${response.status}`);
        }

        return response.json() as Promise<T>;
    }

    async createUser(): Promise<string> {
        const response = await this.request<{ userUID: string }>('/dpdp/users', {
            method: 'POST',
        });
        return response.userUID;
    }

    async storePermission(userUID: string, data: PermissionData): Promise<string> {
        const response = await this.request<{ permissionId: string }>('/dpdp/permissions', {
            method: 'POST',
            body: JSON.stringify({ userUID, data }),
        });
        return response.permissionId;
    }

    async storeData(userUID: string, data: DataRecordInput): Promise<string> {
        const response = await this.request<{ dataId: string }>('/dpdp/data', {
            method: 'POST',
            body: JSON.stringify({ userUID, data }),
        });
        return response.dataId;
    }

    async revokePermission(userUID: string, id: string): Promise<void> {
        await this.request<{ message: string }>('/dpdp/permissions/revoke', {
            method: 'POST',
            body: JSON.stringify({ userUID, permissionId: id }),
        });
    }

    async revokeData(userUID: string, id: string): Promise<void> {
        await this.request<{ message: string }>('/dpdp/data/revoke', {
            method: 'POST',
            body: JSON.stringify({ userUID, dataId: id }),
        });
    }

    async deletePermission(userUID: string, id: string): Promise<void> {
        await this.request<{ message: string }>('/dpdp/permissions/delete', {
            method: 'DELETE',
            body: JSON.stringify({ userUID, permissionId: id }),
        });
    }

    async deleteData(userUID: string, id: string): Promise<void> {
        await this.request<{ message: string }>('/dpdp/data/delete', {
            method: 'DELETE',
            body: JSON.stringify({ userUID, dataId: id }),
        });
    }

    async viewPermission(userUID: string, id: string): Promise<any> {
        return this.request<any>(`/dpdp/permissions/${userUID}/${id}`, {
            method: 'GET',
        });
    }

    async viewData(userUID: string, id: string): Promise<any> {
        return this.request<any>(`/dpdp/data/${userUID}/${id}`, {
            method: 'GET',
        });
    }
}
